test/functional/apm_logging_tests.js

```diff
@@ -0,0 +1,225 @@
+  'serverDescriptionChanged',
+  'serverHeartbeatStarted',
+  'serverHeartbeatSucceeded',
+  'serverHeartbeatFailed',
+  'serverClosed',
+  'topologyOpening',
+  'topologyClosed',
+  'topologyDescriptionChanged'
+];
+
+function stringify(obj) {
+  return EJSON.stringify(obj, { relaxed: true });
+}
+
+function removeTestLogs() {
+  fs
+    .readdirSync(process.cwd())
+    .filter(x => x.indexOf('apm_test_log') !== -1)
+    .forEach(file => fs.unlinkSync(file));
+}
+
+function getFileData(file) {
+  return file
+    .trim()
+    .split('\n')
+    .reduce(
+      (data, current) => {
+        const event = EJSON.parse(current);
+        data.eventStrings.push(stringify(event.object));
+        data[event.type].push(event);
+        data.numLines += 1;
+        return data;
+      },
+      { numLines: 0, command: [], sdam: [], eventStrings: [] }
+    );
+}
+
+function runTestClient(context, urlOptions, closeCb) {
+  const configuration = context.configuration;
+  const url = configuration.url() + urlOptions;
+  const client = new MongoClient(url, { w: 1, useNewUrlParser: true });
+  const events = [];
+
+  loggingEvents.forEach(event => client.on(event, e => events.push(stringify(e))));
+
+  client.connect(function(err) {
+    expect(err).to.be.null;
+    client.close(err => {
+      // set timeout to make sure all file writes are completed
+      setTimeout(() => {
+        closeCb(err, events);
+      }, 10);
+    });
+  });
+}
+
+describe('APM Logging', function() {
+  before(function() {
+    return setupDatabase(this.configuration);
+  });
+
+  after(function() {
+    removeTestLogs();
+  });
+
+  it('should create stream to default location if monitorOut not specified', function(done) {
+    const configuration = this.configuration;
+    const url = configuration.url() + '?monitor=command';
+    const client = new MongoClient(url, { w: 1, useNewUrlParser: true });
+
+    // avoid overwriting any of the user's pre-existing logs at the default location
+    const stub = sinon.stub(fs, 'createWriteStream');
+    const defaultFile = process.cwd() + '/monitor.log';
+
+    client.connect(function() {
+      expect(stub).to.have.been.calledWith(defaultFile);
+      stub.restore();
+      done();
+    });
+  });
+
+  it('should log to stdout if specified', function(done) {
+    const spy = sinon.spy(process.stdout, 'write');
+    const cb = err => {
+      expect(spy).to.have.been.calledTwice;
+      spy.restore();
+      done(err);
+    };
+    runTestClient(this, '?monitor=command&monitorOut=stdout', cb);
+  });
+
+  it('should log to stderr if specified', function(done) {
+    const spy = sinon.spy(process.stderr, 'write');
+    const cb = err => {
+      expect(spy).to.have.been.calledTwice;
+      spy.restore();
+      done(err);
+    };
+    runTestClient(this, '?monitor=command&monitorOut=stderr', cb);
+  });
+
+  it('log only sdam events', function(done) {
+    const cb = err => {
+      const logFile = fs.readFileSync(testLogFile, 'utf8');
```

Comment:
a good rule of thumb for testing is that you generally shouldn't create files unless you absolutely have to. In this case, we can write to e.g. `stderr`/`stdout` or even some sort of virtual stream (through stubbing). We trust in node's abilities to write to a file, so let's just unit test things _around_ that (generation of file names, etc), and when we test the actual output of the logging let's not directly use files